Encoder and Decoder Traits as API

localthomas opened the pull request:
As discussed in jansol/texpresso#3, the `Format` enum is made non-exhaustive, forcing users of the crate to consider the case that an unknown format arises.

localthomas committed:
made Format non-exhaustive for backwards compatibility

jansol commented:
Are there really any uses for this beyond implementing Display? Applications have to manually map between say, DDS format enums and squish formats anyway and it's not really possible to use this crate without specifying a format explicitly since everything is implemented as a trait on the format enum.

localthomas commented:
> Are there really any uses for this beyond implementing Display?

I think it is impossible to predict how this library might be used by applications. But making the enum non-exhaustive makes clear that there might be changes to the enum in the future (especially adding formats).

I would argue that it might be even better to remove the enum entirely, add a trait that provides the current functions, and add `Bc1`, `Bc2`, ... as structs that implement this trait. That would be similar to what the image crate does (with the [`ImageDecoder` trait](https://github.com/image-rs/image/blob/0f1c38a97ec809f399a2a5edfb85c7e17fcd46ec/src/image.rs#L580)). When adding a new format, nothing breaks the API.

jansol commented:
I guess adding Decode and Encode traits and using structs that implement those instead of the enum does make the most sense, yes. That also cleans up the code a bit since there is no need to check `self` anymore.

localthomas commented:
Then I'll change the API to use two traits. I will update this pull request with new commits accordingly.

fintelia commented:
Up to you, but switching from an enum to traits isn't necessarily nicer. Dispatching based on a dyn Trait is going to be slower than based on an enum, and trying to avoid dynamic dispatch might push more of the matching into downstream user code.

localthomas committed:
introduced encoder and decoder traits

The new public traits are sealed to prevent implementations from outside.
The current enum implementations were transferred directly into the new format without changes.

localthomas committed:
refactored internal structure

Moved each BCn implementation into its own module and moved their respective tests with them.

localthomas commented:
I've added two commits that introduce the sealed decoder and encoder traits. This allows a clean separation of BCn formats, so I also refactored their implementations into separate modules (moving the unit tests as well).

As you can see in ` squish_cli/src/main.rs`, this changes the API for consumers that want to use an enumeration. Yes, this is the case with squish_cli, but I argue that this is a bad example. I think such enumerations should be the responsibility of the consumers of the library (this is the case with image and their [`DXTVariant`](https://github.com/image-rs/image/blob/0f1c38a97ec809f399a2a5edfb85c7e17fcd46ec/src/codecs/dxt.rs#L21) enum).

---

> Dispatching based on a dyn Trait [...]

I don't understand what dynamic dispatch you mean. Maybe I'm missing something, but my current implementation of traits does not use something dynamic anywhere. Or do you mean a `match` or `if` block with dynamic dispatch in this case?

> [...] trying to avoid dynamic dispatch might push more of the matching into downstream user code.

I'd argue that there is hardly any difference between calling `Format::Bc1.decompress` or `BC1::decompress`, but the enum could be misused (hence my suggestion of making it non-exhaustive). The selection of which BCn format to use happens in the consumer no matter what. This library can not infer the correct format based on the binary input data. The matching always comes from another source (e.g. a format specifier in a DDS file). Or do I miss something here?

---

I think there is the option to keep an enum, but I would suggest adding `#[non_exhaustive]` and using the trait implementation at least internally (making the traits private) and only exposing `decompress`, `compress` and `compressed_size`. The other option would be roughly what I committed now, with two traits and separate implementations exposed via the API.

@jansol, what do you think?

jansol commented:
I think the aforementioned dynamic dispatch would come from the user doing something to the effect of
```
let format = pick_best_bcn_format(for_a, given, &image);
format.encode(data);
```
if squish exposes a generic Encoder trait that all formats implement. The enum variants don't have this problem since it's always the same type... However it's a design decision the user has to do so I'm not sure that is really something squish needs to care about.

I think implementation-wise having a separate `trait impl` for each probably ends up being more elegant, especially once BC6h and BC7 are added. Unfortunately my own project that I originally did this rewrite for never got far enough to actually make use of this crate so in terms of API preferences I would very much like to hear others' experiences.

I have also been considering forking development off to a new crate to not "taint" the `squish` name and cause confusion regarding the capabilities and quality of the original libsquish (BC4/BC5 is so-so since it's using the exact same algorithm underneath anyway but BC6h and BC7 will be completely different beasts). A different name would feel more appropriate for having different algorithms and experimenting with API more.

Having thought a bit more about it now, I think I'll do the fork later this week unless there are major objections.

jansol commented:
https://github.com/jansol/texpresso is now up, let's continue there.
